File: python/src/_static_disk_index.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT license.

# 本檔案定義了 `StaticDiskIndex` 類別，這是使用者與靜態磁碟索引互動的主要 Python 介面。
# 它封裝了底層的 C++ `PQFlashIndex` 綁定，提供了參數驗證和更易於使用的 API。

import os
import warnings
from typing import Optional

import numpy as np

from . import (
    DistanceMetric,
    QueryResponse,
    QueryResponseBatch,
    VectorDType,
    VectorLike,
    VectorLikeBatch,
)
from . import _diskannpy as _native_dap
from ._common import (
    _assert,
    _assert_2d,
    _assert_is_nonnegative_uint32,
    _assert_is_positive_uint32,
    _castable_dtype_or_raise,
    _ensure_index_metadata,
    _valid_index_prefix,
    _valid_metric,
)

__ALL__ = ["StaticDiskIndex"]


class StaticDiskIndex:
    """
    一個不可變的、以磁碟為後端的 DiskANN 索引。
    """

    def __init__(
        self,
        index_directory: str,
        num_threads: int,
        num_nodes_to_cache: int,
        cache_mechanism: int = 1,
        distance_metric: Optional[DistanceMetric] = None,
        vector_dtype: Optional[VectorDType] = None,
        dimensions: Optional[int] = None,
        index_prefix: str = "ann",
    ):
        """
        建構函式，用於載入一個已存在的磁碟索引。

        ### 參數
        - **index_directory**: 包含索引檔案的目錄。
        - **num_threads**: 搜尋時使用的執行緒數。
        - **num_nodes_to_cache**: 要快取到記憶體中的節點數量。
        - **cache_mechanism**: 快取策略。1=根據樣本查詢快取熱點，2=快取圖的頂層節點。
        - ... (其他元資料參數)
        """
        index_prefix_path = _valid_index_prefix(index_directory, index_prefix)
        
        # --- 載入元資料 ---
        vector_dtype, metric, _, _ = _ensure_index_metadata(
            index_prefix_path,
            vector_dtype,
            distance_metric,
            1,  # 在此上下文中 max_vectors 不重要
            dimensions,
        )
        dap_metric = _valid_metric(metric)

        # --- 參數驗證 ---
        _assert_is_nonnegative_uint32(num_threads, "num_threads")
        _assert_is_nonnegative_uint32(num_nodes_to_cache, "num_nodes_to_cache")

        self._vector_dtype = vector_dtype
        
        # --- 根據資料類型，選擇對應的底層 C++ 原生類別 ---
        if vector_dtype == np.uint8:
            _index = _native_dap.StaticDiskUInt8Index
        elif vector_dtype == np.int8:
            _index = _native_dap.StaticDiskInt8Index
        else:
            _index = _native_dap.StaticDiskFloatIndex
        
        # --- 實例化並載入 C++ 索引物件 ---
        # 這裡會呼叫 C++ 封裝層的建構函式，進而觸發底層 C++ `PQFlashIndex` 的 `load` 方法，
        # 並根據 cache_mechanism 執行預熱。
        self._index = _index(
            distance_metric=dap_metric,
            index_path_prefix=index_prefix_path,
            num_threads=num_threads,
            num_nodes_to_cache=num_nodes_to_cache,
            cache_mechanism=cache_mechanism,
        )

    def search(
        self, query: VectorLike, k_neighbors: int, complexity: int, beam_width: int = 2
    ) -> QueryResponse:
        """
        對單一查詢向量執行搜尋。

        ### 參數
        - **query**: 一維 NumPy 陣列。
        - **k_neighbors**: 要返回的鄰居數量。
        - **complexity**: 搜尋候選集大小 (L)。
        - **beam_width**: 光束寬度，控制每次迭代發出的 I/O 請求數量，是 SSD 索引的關鍵效能參數。
        """
        # --- 搜尋前的參數驗證 ---
        _query = _castable_dtype_or_raise(query, expected=self._vector_dtype)
        _assert(len(_query.shape) == 1, "query vector must be 1-d")
        _assert_is_positive_uint32(k_neighbors, "k_neighbors")
        _assert_is_positive_uint32(complexity, "complexity")
        _assert_is_positive_uint32(beam_width, "beam_width")

        if k_neighbors > complexity:
            warnings.warn(
                f"{k_neighbors=} asked for, but {complexity=} was smaller. Increasing {complexity} to {k_neighbors}"
            )
            complexity = k_neighbors

        # --- 呼叫底層 C++ 搜尋函式 ---
        neighbors, distances = self._index.search(
            query=_query,
            knn=k_neighbors,
            complexity=complexity,
            beam_width=beam_width,
        )
        return QueryResponse(identifiers=neighbors, distances=distances)

    def batch_search(
        self,
        queries: VectorLikeBatch,
        k_neighbors: int,
        complexity: int,
        num_threads: int,
        beam_width: int = 2,
    ) -> QueryResponseBatch:
        """
        對一批查詢向量執行平行的批次搜尋。
        """
        # --- 批次搜尋前的參數驗證 ---
        _queries = _castable_dtype_or_raise(queries, expected=self._vector_dtype)
        _assert_2d(_queries, "queries")
        _assert_is_positive_uint32(k_neighbors, "k_neighbors")
        _assert_is_positive_uint32(complexity, "complexity")
        _assert_is_nonnegative_uint32(num_threads, "num_threads")
        _assert_is_positive_uint32(beam_width, "beam_width")

        if k_neighbors > complexity:
            warnings.warn(
                f"{k_neighbors=} asked for, but {complexity=} was smaller. Increasing {complexity} to {k_neighbors}"
            )
            complexity = k_neighbors

        num_queries, dim = _queries.shape
        neighbors, distances = self._index.batch_search(
            queries=_queries,
            num_queries=num_queries,
            knn=k_neighbors,
            complexity=complexity,
            beam_width=beam_width,
            num_threads=num_threads,
        )
        return QueryResponseBatch(identifiers=neighbors, distances=distances)

File: python/src/static_disk_index.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

// 本檔案實作了 `StaticDiskIndex` 類別的 Python 封裝。
// 這個類別允許 Python 載入一個預先建立好的、為 SSD 優化的磁碟索引 (`PQFlashIndex`)，
// 並對其執行高效能的搜尋操作。

#include "static_disk_index.h"

#include "pybind11/numpy.h"

namespace diskannpy
{

// StaticDiskIndex 的建構函式 (對應 Python 的 __init__)
template <typename DT>
StaticDiskIndex<DT>::StaticDiskIndex(const diskann::Metric metric, const std::string &index_path_prefix,
                                     const uint32_t num_threads, const size_t num_nodes_to_cache,
                                     const uint32_t cache_mechanism)
    // 建立一個平台相關的對齊檔案讀取器和一個 PQFlashIndex 物件
    : _reader(std::make_shared<PlatformSpecificAlignedFileReader>()), _index(_reader, metric)
{
    const uint32_t _num_threads = num_threads != 0 ? num_threads : omp_get_num_procs();
    
    // 呼叫底層 C++ PQFlashIndex 物件的 load 方法來載入磁碟索引檔案。
    int load_success = _index.load(_num_threads, index_path_prefix.c_str());
    if (load_success != 0)
    {
        throw std::runtime_error("index load failed.");
    }

    // 根據使用者選擇的快取機制，預先載入部分節點到記憶體中以「預熱」索引。
    if (cache_mechanism == 1)
    {
        // 策略 1: 根據樣本查詢，快取最常被訪問的「熱點」節點。
        std::string sample_file = index_path_prefix + std::string("_sample_data.bin");
        cache_sample_paths(num_nodes_to_cache, sample_file, _num_threads);
    }
    else if (cache_mechanism == 2)
    {
        // 策略 2: 快取圖的頂部幾層節點 (BFS)。
        cache_bfs_levels(num_nodes_to_cache);
    }
}

// 快取圖的頂部幾層節點 (BFS 策略)
template <typename DT> void StaticDiskIndex<DT>::cache_bfs_levels(const size_t num_nodes_to_cache)
{
    std::vector<uint32_t> node_list;
    _index.cache_bfs_levels(num_nodes_to_cache, node_list);
    _index.load_cache_list(node_list);
}

// 根據樣本查詢路徑來快取「熱點」節點
template <typename DT>
void StaticDiskIndex<DT>::cache_sample_paths(const size_t num_nodes_to_cache, const std::string &warmup_query_file,
                                             const uint32_t num_threads)
{
    if (!file_exists(warmup_query_file))
    {
        return;
    }

    std::vector<uint32_t> node_list;
    // 執行樣本查詢以產生最常訪問的節點列表
    _index.generate_cache_list_from_sample_queries(warmup_query_file, 15, 4, num_nodes_to_cache, num_threads,
                                                   node_list);
    // 載入這些節點到快取中
    _index.load_cache_list(node_list);
}

// 單一查詢搜尋
template <typename DT>
NeighborsAndDistances<StaticIdType> StaticDiskIndex<DT>::search(
    py::array_t<DT, py::array::c_style | py::array::forcecast> &query, const uint64_t knn, const uint64_t complexity,
    const uint64_t beam_width)
{
    py::array_t<StaticIdType> ids(knn);
    py::array_t<float> dists(knn);

    std::vector<uint32_t> u32_ids(knn);
    std::vector<uint64_t> u64_ids(knn);
    diskann::QueryStats stats;

    // 呼叫底層 C++ PQFlashIndex 的核心搜尋函式 `cached_beam_search`
    // complexity: 搜尋候選集大小 (L)
    // beam_width: 光束寬度，影響 I/O 數量和搜尋廣度
    _index.cached_beam_search(query.data(), knn, complexity, u64_ids.data(), dists.mutable_data(), beam_width, false,
                              &stats);

    // 將 C++ 返回的 uint64_t 結果複製到 Python 的 uint32_t NumPy 陣列中
    auto r = ids.mutable_unchecked<1>();
    for (uint64_t i = 0; i < knn; ++i)
        r(i) = (unsigned)u64_ids[i];

    return std::make_pair(ids, dists);
}

// 批次搜尋
template <typename DT>
NeighborsAndDistances<StaticIdType> StaticDiskIndex<DT>::batch_search(
    py::array_t<DT, py::array::c_style | py::array::forcecast> &queries, const uint64_t num_queries, const uint64_t knn,
    const uint64_t complexity, const uint64_t beam_width, const uint32_t num_threads)
{
    py::array_t<StaticIdType> ids({num_queries, knn});
    py::array_t<float> dists({num_queries, knn});

    omp_set_num_threads(num_threads);

    std::vector<uint64_t> u64_ids(knn * num_queries);

#pragma omp parallel for schedule(dynamic, 1) default(none)                                                            \
    shared(num_queries, queries, knn, complexity, u64_ids, dists, beam_width)
    for (int64_t i = 0; i < (int64_t)num_queries; i++)
    {
        _index.cached_beam_search(queries.data(i), knn, complexity, u64_ids.data() + i * knn, dists.mutable_data(i),
                                  beam_width);
    }

    // 將結果從 1D 的 C++ 向量複製到 2D 的 Python NumPy 陣列
    auto r = ids.mutable_unchecked();
    for (uint64_t i = 0; i < num_queries; ++i)
        for (uint64_t j = 0; j < knn; ++j)
            r(i, j) = (uint32_t)u64_ids[i * knn + j];

    return std::make_pair(ids, dists);
}

// 模板實例化
template class StaticDiskIndex<float>;
template class StaticDiskIndex<uint8_t>;
template class StaticDiskIndex<int8_t>;
} // namespace diskannpy
